### PR TITLE
[boschshc] Handle relay mode changes during initialization

### LINKED ISSUE #16590: [boschshc] Add support for Relay - 'MICROMODULE_RELAY'
This request is to add support for another Bosch Smart Home device, namely Relay:

From openhab.log:
```Unknown deviceModel 'MICROMODULE_RELAY'! Please create a support request issue for this unknown device model.```

```
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - Found device: name=Relais Test id=hdm:ZigBee:30XXXXXXXXXXXXXX
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: CommunicationQuality
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: PowerSwitch
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: ChildProtection
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: PowerSwitchProgram
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: ElectricalFaults
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: SwitchConfiguration
2024-03-28 [DEBUG] [rnal.discovery.ThingDiscoveryService] - .... service: Linking
```

```
{"@type":"device","rootDeviceId":"64-da-a0-10-83-a7","id":"hdm:ZigBee:30fb10fffe46d732","deviceServiceIds":["CommunicationQuality","PowerSwitch","ChildProtection","PowerSwitchProgram","ElectricalFaults","SwitchConfiguration","Linking"],"manufacturer":"BOSCH","roomId":"hz_16","deviceModel":"MICROMODULE_RELAY","serial":"30FB10FFFE46D732","profile":"GENERIC","iconId":"ip_heat_pump","name":"Relais Test","status":"AVAILABLE","childDeviceIds":[],"supportedProfiles":["LIGHT","GENERIC","HEATING_RCC"]}
```

Please let me know if I'd need to provide further details or can support testing.

Happy Easter!

Regards,
Michael

### PR DESCRIPTION
**Pull request by david-pace**
This is an addition to https://github.com/openhab/openhab-addons/pull/17026, which was merged too early.

If the relay mode (power switch / impulse switch) is changed (which happens rarely, e.g. when the device is repurposed), the thing status changes to `CONFIGURATION_ERROR`. The user is prompted to re-create the thing.

Additionally, a property is set during the initialization that signals the current relay mode.

Includes an enhancement for the `impulse-length` channel to accept `QuantityType` commands in addition to `DecimalType` commands.

~~**Note:** This PR has to be tested by @mike-bike with the actual hardware before it can be merged.~~

closes #16590

### CODE REVIEW

File: bundles/org.openhab.binding.boschshc/README.md [hunk 1]
```diff
@@ -238,6 +238,8 @@ The smart switching relay is your universal all-rounder for smart switching.
 | impulse-length          | Number:Time | &#9745;  | Channel to configure how long the relay will stay on after receiving an impulse switch event. The time is specified in tenth seconds (deciseconds), e.g. 15 means 1.5 seconds. Only available if the relay is in impulse switch mode. |
 | instant-of-last-impulse | DateTime    | &#9744;  | Timestamp indicating when the last impulse was triggered. Only available if the relay is in impulse switch mode. |
 
+If the device mode is changed from power switch to impulse switch mode or vice versa, the corresponding thing has to be deleted and re-added in OpenHAB.
```
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/README.md:241]:
```suggestion
If the device mode is changed from power switch to impulse switch mode or vice versa, the corresponding thing has to be deleted and re-added in openHAB.
```

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/relay/RelayHandler.java [hunk 2]
```diff
@@ -108,10 +110,26 @@ public RelayHandler(Thing thing) {
     @Override
     protected boolean processDeviceInfo(Device deviceInfo) {
         this.isInImpulseSwitchMode = isRelayInImpulseSwitchMode(deviceInfo);
-        configureChannels();
+        boolean isChannelConfigurationValid = configureChannels();
+        if (!isChannelConfigurationValid)
+            return false;
```
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/relay/RelayHandler.java:115]:
```suggestion
        if (!isChannelConfigurationValid) {
            return false;
        }
```

File: bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/relay/RelayHandler.java [hunk 3]
```diff
@@ -108,10 +110,26 @@ public RelayHandler(Thing thing) {
...
+    private void updateModePropertyIfApplicable() {
+        String modePropertyValue = isInImpulseSwitchMode ? ImpulseSwitchService.IMPULSE_SWITCH_SERVICE_NAME
+                : PowerSwitchService.POWER_SWITCH_SERVICE_NAME;
+
+        Map<String, String> properties = getThing().getProperties();
+        if (properties.containsKey(PROPERTY_MODE) && properties.get(PROPERTY_MODE).equals(modePropertyValue)) {
+            return;
+        }
+
+        updateThing(editThing().withProperty(PROPERTY_MODE, modePropertyValue).build());
```
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/java/org/openhab/binding/boschshc/internal/devices/relay/RelayHandler.java:130]: Could be simplified:
```suggestion
        updateProperty(PROPERTY_MODE, modePropertyValue);
```

File: bundles/org.openhab.binding.boschshc/src/main/resources/OH-INF/i18n/boschshc.properties [hunk 4]
```diff
@@ -220,3 +220,4 @@ offline.conf-error.invalid-device-id = Device ID is invalid.
 offline.conf-error.empty-state-id = No ID set.
 offline.conf-error.invalid-state-id = ID is invalid.
 offline.conf-error.child-device-ids-not-obtainable = Could not obtain child device IDs.
+offline.conf-error.relay-recreation-required = Relay mode (power/impulse switch) change detected. Please delete and re-create this thing.
```
Comment (jlaur) [inline bundles/org.openhab.binding.boschshc/src/main/resources/OH-INF/i18n/boschshc.properties:223]:
```suggestion
offline.conf-error.relay-recreation-required = Relay mode (power/impulse switch) change detected. Please delete and re-create this Thing.
```

**Review by jlaur — COMMENTED** [review]
Thanks for improving the relay support. I have added a few minor comments.

**Review by lsiepel — APPROVED** [review]
Thanks, very clean. LGTM

**Pull request merged**